recommender/services/synthetic_dataset/service_engagement.py

```diff
@@ -5,17 +5,54 @@
 import numpy as np
 import torch
 import pandas as pd
-from torch.utils.tensorboard import SummaryWriter
 
-from definitions import LOG_DIR
-from recommender.engine.agents.rl_agent.utils import get_service_indices, cfr
+from recommender.engine.agents.rl_agent.utils import get_service_indices
 from recommender.models import User, Service
 
 
-def _dist_score(x):
-    """Returns score for a given euclidean distance, between 0 and 1.
+def _distance_metric(cluster: torch.Tensor, point: torch.Tensor) -> float:
+    """Computes euclidean distance between the cluster centroid and
+     a given point and the converts it to a score, between 0 and 1, assuming the cluster and the point are
+     normalized in every dimension, between 0 and 0.5.
     The lesser the distance the better the score"""
-    return -x + 1
+    centroid = cluster.mean(dim=0).unsqueeze(0)
+    point = point.unsqueeze(0)
+    dist = torch.cdist(centroid, point).item()
+    return -dist + 1
+
+
+def _overlap_metric(x: int) -> float:
+    """Returns measure of overlapping niches.
+    Uses a sigmoid function that is shifted to the left"""
+    return 1 / (1 + np.exp(-(x - 1)))
```

Comment:
This is on purpose. The value of this metric at x = 0 is ~ 0.25, which is later fed to the synthetic reward generation, which in turn has the ability to generate a reward other than "exit" (this would not happen if the computed engagement is 0). This allows us to achieve a mirror the level of unpredictability seen in real users - even if the user potentially does not like the service at all, they might still click it.